File: idea-plugin/src/main/java/com/zzf/rikki/idea/agent/compat/CommandRunner.java
package com.zzf.rikki.idea.agent.compat;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

public class CommandRunner {
    public static final long DEFAULT_TIMEOUT_MS = 60_000L;
    private static final int MAX_OUTPUT = 8_000;
    private static final long POLL_INTERVAL_MS = 200L;
    private static final Set<String> SUPPORTED_SHELLS = Set.of("auto", "bash", "powershell", "cmd");

    public CommandRunnerResult run(
            String command,
            String workspaceRoot,
            String workdir,
            long timeoutMs,
            String shellPref,
            AtomicBoolean skipFlag
    ) {
        String normalizedShell = shellPref == null || shellPref.isBlank() ? "auto" : shellPref.trim().toLowerCase();
        if (!SUPPORTED_SHELLS.contains(normalizedShell)) {
            throw new IllegalArgumentException("shell must be one of: auto, bash, powershell, cmd");
        }
        File resolvedWorkdir = resolveWorkdir(workspaceRoot, workdir);
        List<ShellSpec> candidates = resolveShellCandidates(normalizedShell);
        List<String> startupErrors = new ArrayList<>();
        for (ShellSpec shell : candidates) {
            ProcessBuilder processBuilder = new ProcessBuilder();
            List<String> commandLine = new ArrayList<>(shell.argvPrefix);
            commandLine.add(command);
            processBuilder.command(commandLine);
            processBuilder.directory(resolvedWorkdir);
            processBuilder.redirectErrorStream(true);
            Map<String, String> environment = processBuilder.environment();
            environment.putAll(System.getenv());
            environment.put("PYTHONIOENCODING", "utf-8");
            if ("bash".equals(shell.name)) {
                environment.put("LANG", "en_US.UTF-8");
                environment.put("LC_ALL", "en_US.UTF-8");
            }
            Process process;
            try {
                process = processBuilder.start();
            } catch (Exception e) {
                startupErrors.add(shell.name + ": " + (e.getMessage() == null ? "failed to start" : e.getMessage()));
                continue;
            }
            return executeProcess(process, timeoutMs, skipFlag, shell.name);
        }
        String shellNames = String.join(", ", candidates.stream().map(spec -> spec.name).toList());
        String detail = startupErrors.isEmpty() ? "no startup error details" : String.join("; ", startupErrors);
        return new CommandRunnerResult(
                "No usable shell found (requested=" + normalizedShell + ", tried=" + shellNames + "): " + detail,
                -1,
                false,
                false,
                normalizedShell
        );
    }

    public String formatOutput(String command, long timeoutMs, CommandRunnerResult result) {
        String prefix = "[shell=" + result.getShell() + "]";
        String output = result.getOutput();
        String truncated = output.length() > MAX_OUTPUT ? output.substring(0, MAX_OUTPUT) + "\n\n...(truncated)" : output;
        if (result.getSkipped()) {
            return (prefix + " (Skipped by user - command was interrupted)\n" + truncated).trim();
        }
        if (result.getTimedOut()) {
            return (prefix + " (Command timed out after " + timeoutMs + "ms)\n" + truncated).trim();
        }
        if (result.getExitCode() != 0) {
            return (prefix + " Command failed with exit code " + result.getExitCode() + ": " + command + "\n" + truncated).trim();
        }
        if (truncated.isBlank()) {
            return prefix + " (no output)";
        }
        return (prefix + "\n" + truncated).trim();
    }

    private File resolveWorkdir(String workspaceRoot, String requested) {
        File dir = new File(requested == null || requested.isBlank() ? workspaceRoot : requested);
        if (!dir.isAbsolute()) {
            dir = new File(workspaceRoot, requested == null ? "" : requested);
        }
        return dir.isDirectory() ? dir : new File(workspaceRoot);
    }

    private List<ShellSpec> resolveShellCandidates(String shellPref) {
        return switch (shellPref) {
            case "bash" -> List.of(new ShellSpec("bash", List.of("bash", "-c")));
            case "powershell" -> List.of(new ShellSpec("powershell", List.of("powershell", "-NoProfile", "-NonInteractive", "-Command")));
            case "cmd" -> List.of(new ShellSpec("cmd", List.of("cmd", "/c")));
            default -> List.of(
                    new ShellSpec("bash", List.of("bash", "-c")),
                    new ShellSpec("powershell", List.of("powershell", "-NoProfile", "-NonInteractive", "-Command")),
                    new ShellSpec("cmd", List.of("cmd", "/c"))
            );
        };
    }

    private CommandRunnerResult executeProcess(Process process, long timeoutMs, AtomicBoolean skipFlag, String shell) {
        ByteArrayOutputStream outputBuffer = new ByteArrayOutputStream();
        Thread reader = new Thread(() -> {
            try {
                process.getInputStream().transferTo(outputBuffer);
            } catch (Exception ignored) {
            }
        }, "rikki-command-reader");
        reader.setDaemon(true);
        reader.start();

        long deadline = System.currentTimeMillis() + timeoutMs;
        boolean timedOut = false;
        boolean skipped = false;
        while (process.isAlive()) {
            if (skipFlag != null && skipFlag.get()) {
                skipped = true;
                process.destroyForcibly();
                break;
            }
            if (System.currentTimeMillis() >= deadline) {
                timedOut = true;
                process.destroyForcibly();
                break;
            }
            try {
                process.waitFor(POLL_INTERVAL_MS, TimeUnit.MILLISECONDS);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                skipped = true;
                process.destroyForcibly();
                break;
            }
        }
        try {
            reader.join(2_000L);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        int exitCode;
        try {
            exitCode = process.exitValue();
        } catch (Exception ignored) {
            exitCode = -1;
        }
        return new CommandRunnerResult(outputBuffer.toString(java.nio.charset.StandardCharsets.UTF_8), exitCode, timedOut, skipped, shell);
    }

    private record ShellSpec(String name, List<String> argvPrefix) {
    }
}

File: idea-plugin/src/main/kotlin/com/zzf/rikki/idea/agent/compat/CommandRunner.kt
package com.zzf.rikki.idea.agent.compat

import java.io.ByteArrayOutputStream
import java.io.File
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

data class CommandRunnerResult(
    val output: String,
    val exitCode: Int,
    val timedOut: Boolean,
    val skipped: Boolean,
    val shell: String
)

class CommandRunner {
    companion object {
        const val DEFAULT_TIMEOUT_MS = 60_000L
        private const val MAX_OUTPUT = 8_000
        private const val POLL_INTERVAL_MS = 200L
        private val SUPPORTED_SHELLS = setOf("auto", "bash", "powershell", "cmd")
    }

    private data class ShellSpec(val name: String, val argvPrefix: List<String>)

    fun run(
        command: String,
        workspaceRoot: String,
        workdir: String,
        timeoutMs: Long,
        shellPref: String,
        skipFlag: AtomicBoolean? = null
    ): CommandRunnerResult {
        val normalizedShell = shellPref.trim().lowercase().ifBlank { "auto" }
        require(normalizedShell in SUPPORTED_SHELLS) {
            "shell must be one of: auto, bash, powershell, cmd"
        }
        val resolvedWorkdir = resolveWorkdir(workspaceRoot, workdir)
        val candidates = resolveShellCandidates(normalizedShell)
        val startupErrors = mutableListOf<String>()
        for (shell in candidates) {
            val pb = ProcessBuilder(shell.argvPrefix + command)
                .directory(resolvedWorkdir)
                .redirectErrorStream(true)
            pb.environment().putAll(System.getenv())
            pb.environment()["PYTHONIOENCODING"] = "utf-8"
            if (shell.name == "bash") {
                pb.environment()["LANG"] = "en_US.UTF-8"
                pb.environment()["LC_ALL"] = "en_US.UTF-8"
            }
            val process = try {
                pb.start()
            } catch (e: Exception) {
                startupErrors += "${shell.name}: ${e.message ?: "failed to start"}"
                continue
            }
            return executeProcess(process, timeoutMs, skipFlag, shell.name)
        }
        val shellNames = candidates.joinToString(", ") { it.name }
        val detail = if (startupErrors.isEmpty()) {
            "no startup error details"
        } else {
            startupErrors.joinToString("; ")
        }
        return CommandRunnerResult(
            output = "No usable shell found (requested=$normalizedShell, tried=$shellNames): $detail",
            exitCode = -1,
            timedOut = false,
            skipped = false,
            shell = normalizedShell
        )
    }

    fun formatOutput(
        command: String,
        timeoutMs: Long,
        result: CommandRunnerResult
    ): String {
        val prefix = "[shell=${result.shell}]"
        val truncated = if (result.output.length > MAX_OUTPUT) {
            result.output.take(MAX_OUTPUT) + "\n\n...(truncated)"
        } else {
            result.output
        }
        return when {
            result.skipped ->
                "$prefix (Skipped by user - command was interrupted)\n$truncated".trimEnd()

            result.timedOut ->
                "$prefix (Command timed out after ${timeoutMs}ms)\n$truncated".trimEnd()

            result.exitCode != 0 ->
                "$prefix Command failed with exit code ${result.exitCode}: $command\n$truncated".trimEnd()

            truncated.isBlank() ->
                "$prefix (no output)"

            else ->
                "$prefix\n$truncated"
        }
    }

    private fun resolveWorkdir(workspaceRoot: String, requested: String): File {
        val dir = File(requested).let { if (it.isAbsolute) it else File(workspaceRoot, requested) }
        return if (dir.isDirectory) dir else File(workspaceRoot)
    }

    private fun resolveShellCandidates(shellPref: String): List<ShellSpec> = when (shellPref) {
        "bash" -> listOf(ShellSpec("bash", listOf("bash", "-c")))
        "powershell" -> listOf(
            ShellSpec("powershell", listOf("powershell", "-NoProfile", "-NonInteractive", "-Command"))
        )

        "cmd" -> listOf(ShellSpec("cmd", listOf("cmd", "/c")))
        else -> listOf(
            ShellSpec("bash", listOf("bash", "-c")),
            ShellSpec("powershell", listOf("powershell", "-NoProfile", "-NonInteractive", "-Command")),
            ShellSpec("cmd", listOf("cmd", "/c"))
        )
    }

    private fun executeProcess(
        process: Process,
        timeoutMs: Long,
        skipFlag: AtomicBoolean?,
        shell: String
    ): CommandRunnerResult {
        val outputBuf = ByteArrayOutputStream()
        val readerThread = Thread {
            try {
                process.inputStream.copyTo(outputBuf)
            } catch (_: Exception) {
            }
        }.also {
            it.isDaemon = true
            it.start()
        }
        val deadline = System.currentTimeMillis() + timeoutMs
        var timedOut = false
        var skipped = false
        while (process.isAlive) {
            if (skipFlag?.get() == true) {
                skipped = true
                process.destroyForcibly()
                break
            }
            if (System.currentTimeMillis() >= deadline) {
                timedOut = true
                process.destroyForcibly()
                break
            }
            process.waitFor(POLL_INTERVAL_MS, TimeUnit.MILLISECONDS)
        }
        readerThread.join(2_000)
        val output = outputBuf.toString(Charsets.UTF_8.name())
        val exitCode = try {
            process.exitValue()
        } catch (_: Exception) {
            -1
        }
        return CommandRunnerResult(output, exitCode, timedOut, skipped, shell)
    }
}